crates/sp-executor/src/lib.rs

```diff
@@ -19,27 +19,81 @@
 
 use parity_scale_codec::{Decode, Encode};
 use scale_info::TypeInfo;
-use sp_runtime::traits::{Block as BlockT, Hash as HashT, Header as HeaderT};
+use sp_core::H256;
+use sp_runtime::traits::{BlakeTwo256, Block as BlockT, Hash as HashT, Header as HeaderT};
 use sp_runtime::{OpaqueExtrinsic, RuntimeDebug};
 use sp_std::vec::Vec;
 use sp_trie::StorageProof;
 
-/// Dummy bundle header.
-pub type BundleHeader = Vec<u8>;
+/// Header of transaction bundle.
+#[derive(Decode, Encode, TypeInfo, PartialEq, Eq, Clone, RuntimeDebug)]
+pub struct BundleHeader {
+    /// The slot number.
+    pub slot_number: u64,
+    /// The merkle root of the extrinsics.
+    pub extrinsics_root: H256,
+}
+
+impl BundleHeader {
+    /// Returns the hash of this header.
+    pub fn hash(&self) -> H256 {
+        BlakeTwo256::hash_of(self)
+    }
+}
 
 /// Transaction bundle
 #[derive(Decode, Encode, TypeInfo, PartialEq, Eq, Clone, RuntimeDebug)]
-pub struct Bundle {
+pub struct Bundle<Extrinsic> {
     /// The bundle header.
     pub header: BundleHeader,
     /// THe accompanying extrinsics.
-    pub opaque_transactions: Vec<u8>,
+    pub extrinsics: Vec<Extrinsic>,
+}
+
+impl<Extrinsic> Bundle<Extrinsic> {
+    /// Returns the hash of this bundle.
+    pub fn hash(&self) -> H256 {
+        self.header.hash()
+    }
+}
+
+/// Encoded extrinsic.
+#[derive(Decode, Encode, TypeInfo, PartialEq, Eq, Clone, RuntimeDebug)]
+pub struct EncodedExtrinsic(Vec<u8>);
```

Comment:
There is `sp_runtime::OpaqueExtrinsic`, which is exactly like this. Any specific reason to not use it?

Reply:
See https://github.com/paritytech/substrate/pull/10504. But After this has been merged, I realized that `EncodedExtrinsic` is more straightforward because using `OpaqueExtrinsic` you'll have to deal with `Result` using `expect()`, `EncodeExtrinsic` could be just clearer.

Reply:
OK, I had another PR merged a few days ago that would be useful to get in, maybe I can just go ahead and upgrade substrate to the latest state of upstream's `master`, WDYT?

Also would be nice to have a TODO with link to an upstream PR explaining that this is a temporary measure.

Reply:
I already have a branch with Substrate upgrade and switch to `OpaqueExtrinsic`

Reply:
>OK, I had another PR merged a few days ago that would be useful to get in, maybe I can just go ahead and upgrade substrate to the latest state of upstream's master, WDYT?

Yeah, it's totally fine to use the latest master of substrate IMO.

>Also would be nice to have a TODO with link to an upstream PR explaining that this is a temporary measure.

That's what I usually would do. The reason for not doing it here is that now I don't fully sure that `OpaqueExtrinsic` is better than `EncodedExtrinsic` in this sutiation because you have to use `expect` in https://github.com/subspace/subspace/pull/207/files#diff-c49314f00f4e4f47d787bfe52ec77eef8dcbd2e37fab3357910b86ff4cd49d5aR83 . So the small difference between `OpaqueExtrinsic` and `EncodedExtrinsic` is that `EncodedExtrinsic` implies the source of input but `OpaqueExtrinsic` assumes any bytes leading to the fallible result and you have to deal with that. No big deal, just `EncodedExtrinsic` can be more specific and cleaner.

Reply:
I was actually thinking it might make sense to do `impl<E: Extrinsic> From<E> for OpaqueExtrinsic` upstream, since it will, obviously, always be correct, it will also allow to avoid one memory allocation (no need to copy `Vec` into another `Vec`), do you think this is a good idea?

Reply:
Make sense to me, I was expecting there is a more convenient way to convert `Extrinsic` to `OpaqueExtrinsic`.